### DISCUSSION TITLE
TailwindCSS JIT crash assistance.

### DISCUSSION THREAD

**Discussion by ariesclark** [tailwindlabs/tailwindcss#4919] [Help]
I cannot figure out what's been causing this error. I haven't been able to make a reproduction properly nor have I been able to fix this; the error appeared randomly and hasn't left.

It only happens with JIT mode enabled.

### Console
```
> next dev

ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Using webpack 4. Reason: webpack5 flag is set to false in next.config.js https://nextjs.org/docs/messages/webpack5
Warning: A null PostCSS plugin was provided. This entry will be ignored.
Warning: A null PostCSS plugin was provided. This entry will be ignored.

warn - You have enabled the JIT engine which is currently in preview.
warn - Preview features are not covered by semver, may introduce breaking changes, and can change at any time.
error - ./src/styles/root.css ((webpack)/css-loader/cjs.js??ref--5-oneOf-6-1!(webpack)/postcss-loader/cjs.js??ref--5-oneOf-6-2!./src/styles/root.css)
RangeError: Maximum call stack size exceeded
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
```

Slightly different error with Next.js Webpack 5 enabled.
```
> next dev

ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
Warning: A null PostCSS plugin was provided. This entry will be ignored.
Warning: A null PostCSS plugin was provided. This entry will be ignored.

warn - You have enabled the JIT engine which is currently in preview.
warn - Preview features are not covered by semver, may introduce breaking changes, and can change at any time.
error - ./node_modules/.pnpm/next@11.0.1_f14f4a1115575f7d598b350caa6c0011/node_modules/next/dist/compiled/css-loader/cjs.js??ruleSet[1].rules[2].oneOf[6].use[1]!./node_modules/.pnpm/next@11.0.1_f14f4a1115575f7d598b350caa6c0011/node_modules/next/dist/compiled/postcss-loader/cjs.js??ruleSet[1].rules[2].oneOf[6].use[2]!./src/styles/root.css
RangeError: Maximum call stack size exceeded
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
    at candidatePermutations.next (<anonymous>)
```

### tailwind.config.js
```js
/* eslint-disable no-undef */

const colors = require("./src/styles/colors");

module.exports = {
    mode: "jit",
    purge: [
        "./src/**/*.{ts,tsx}",
    ]
};
```

### postcss.config.js
```js
/* eslint-disable no-undef */

//const colors = require("./src/styles/colors");
const isDevelopment = process.env["NODE_ENV"] === "development";

module.exports = {
    plugins: [
        ["postcss-import", {}],
        ["tailwindcss/nesting", {}],
        ["tailwindcss", {}],
        ["autoprefixer", {}],
        isDevelopment ? null : ["cssnano", {}]
    ]
};
```

### root.css
```css
@tailwind base;
@tailwind components;
@tailwind utilities;
```

pnpx tailwindcss -c tailwind.config.js -m -o src/styles/dist/tailwind.css

```
tailwindcss/lib/jit/lib/generateRules.js:37
function* candidatePermutations(candidate, lastIndex = Infinity) {
                               ^

RangeError: Maximum call stack size exceeded
```
tried using the tailwindcss cli.

Any help would be lovely!

**Comment by adamwathan** [comment]
Hey! Can you provide a reproduction repository? I don't see anything wrong at a glance in the stuff you've pasted but the one thing that's missing is your template file content, and that's likely what would be triggering this issue. It looks like an infinite recursion problem, so it's probably a specific string in a template that we aren't handling intelligently enough.

**Comment by madelynschuste** [comment]
Just to make sure: the concerns about >10MB CSS files and expensive configs only really affect compilation times, right? I always quite liked the simplicity of purging every class that has no literal occurrence anywhere in your code, i.e. a simple search rather than having to parse or compile my code.

And if I understand it correctly, this would just be the other way around: rather than starting with a big stylesheet and purging classes that do not literally occur, it starts with an empty stylesheet and adds the classes that do occur?

[MyAARPMedicare Login](https://www.myaarpmedicare.work/)

**Reply by simonswiss** [reply]
Hey @madelynschuste,

Yep, in a nutshell, that's how the JIT engine works. You start with a very barebone CSS file (basically the `base` layer), and as you start using utility classes in your templates, they get added to the generated CSS output.

For large files without JIT, the concerns are indeed around compilation time. But also the dev tools experience: At a certain threshold, the dev tools (Chrome dev tools for example) get very laggy and struggle a bit.
